src/backend/cuda/cudnn.hpp

```diff
@@ -10,22 +10,136 @@
-#include <cudnn.h>
+#include <cudnnModule.hpp>
 
 namespace cuda {
 
 const char *errorString(cudnnStatus_t err);
 
-#define CUDNN_CHECK(fn)                                                    \
-    do {                                                                   \
-        cudnnStatus_t _error = (fn);                                       \
-        if (_error != CUDNN_STATUS_SUCCESS) {                              \
-            char _err_msg[1024];                                           \
-            snprintf(_err_msg, sizeof(_err_msg), "CUDNN Error (%d): %s\n", \
-                     (int)(_error), errorString(_error));                  \
-                                                                           \
-            AF_ERROR(_err_msg, AF_ERR_INTERNAL);                           \
-        }                                                                  \
+#define CUDNN_CHECK(fn)                                                     \
+    do {                                                                    \
+        cudnnStatus_t _error = (fn);                                        \
+        if (_error == CUDNN_STATUS_SUCCESS) {                               \
+            break;                                                          \
+        } else if (_error == CUDNN_STATUS_ALLOC_FAILED) {                   \
+            AF_ERROR(                                                       \
+                "CUDNN Error(CUDNN_STATUS_ALLOC_FAILED): Error allocating " \
+                "for function all ",                                        \
+                AF_ERR_NO_MEM);                                             \
+        } else if (_error == CUDNN_STATUS_NOT_SUPPORTED) {                  \
+            CUDA_NOT_SUPPORTED(                                             \
+                "CUDNN Error(CUDNN_STATUS_NOT_SUPPORTED): This version of " \
+                "CUDNN does not support the data type or the size of this " \
+                "operation");                                               \
+        } else {                                                            \
+            char _err_msg[1024];                                            \
+            snprintf(_err_msg, sizeof(_err_msg), "CUDNN Error(%s): \n",     \
+                     errorString(_error));                                  \
+            AF_ERROR(_err_msg, AF_ERR_INTERNAL);                            \
+        }                                                                   \
     } while (0)
 
-}
+
+
+// cuDNN Wrappers
+//
+// cuDNN deprecates and releases function names often between releases. in order
+// to prevent locking arrayfire versions to specific cuDNN versions, we wrap all
+// cuDNN calls so that the main codebase is not full of ifdefs. The Following
+// functions are wrappers around cuDNN functions that abstract out the version
+// differences between older versions of cuDNN.
+//
```

Comment:
How do we handle this inside the wrapper call or the cuDNN module loader ? There doesn't seem to be any guards inside cuDNN module loader based on the cuDNN versions.